### PR TITLE
BFD-1921: Provide a library for npi org lookups

### PR DESCRIPTION
**Pull request by cbrunefearless**
**JIRA Ticket:**
[BFD-1921](https://jira.cms.gov/browse/BFD-1921)

**User Story or Bug Summary:**

<p style="margin: 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><b>User Story</b><br>As a developer, I want to add a new maven library independent of the current NPI Org utility so I can use it as a standalone library. This new library will able to take the directory where the file will downloaded and copy it to the resource directory. The library will also include the NPIOrgDisplayLookup to lookup drug codes.</p><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><b>Original File Output</b></p><div class="table-wrap" style="margin: 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

NPI | Entity Type Code | Replacement NPI | Employer Identification Number (EIN) | Provider Organization Name (Legal Business Name) | Provider Last Name (Legal Name) | Provider First Name
-- | -- | -- | -- | -- | -- | --
1497758544 | 2 |   | <UNAVAIL> | CUMBERLAND COUNTY HOSPITAL SYSTEM, INC |  
1215930367 | 1 |   |   |   | GRESSOT | LAURENT
1023011178 | 2 |   | <UNAVAIL> | COLLABRIA CARE |  
1932102084 | 1 |   |   |   | ADUSUMILLI | RAVI
1841293990 | 2 |   |   |   | WORTSMAN | SUSAN
1750384806 | 3 |   |   |   | BISBEE | ROBERT
1669475711 | 4 |   |   |   | SUNG | BIN

</div><p style="margin: 10px 0px 0px; padding: 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><b>AC</b></p><ol style="margin: 10px 0px 0px; color: rgb(23, 43, 77); font-family: -apple-system, &quot;system-ui&quot;, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li>A new maven project is added to the apps directory<ol style="margin: 0px;"><li>produce a jar file artifact that exposes a lookup class for NPI Org code that matches existing functionality</li><li>the jar file should contain all necessary data (e.g., csv file)</li></ol></li><li>Don't include the new maven module in the bfd-parent's modules declaration, but should be a child of bfd-parent<ol style="margin: 0px;"><li>Something in GH Actions should run `mvn verify` on this new module, to verify that it works and tests pass when it's modified.</li></ol></li><li>Lookups are performed against data transformed as above (individual provider NPIs will not be present)</li><li>Research which entries are valid org NPIs and only include those.</li></ol>

---

### What Does This PR Do?

Adds a jar to our current project that pulls data for NPI organization data.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-data-npi/build.sh [hunk 1]
```diff
@@ -0,0 +1,5 @@
+mvn --quiet clean install
+
+mvn --quiet exec:java \
+-Dexec.mainClass="gov.cms.bfd.data.npi.utility.App" \
+-Dexec.args="./src/main/resources/"
```
Comment (christrotteradhoc) [inline apps/bfd-data-npi/build.sh:5]: lets remove this script and instead add support via system property to the maven build for optionally providing the location of the NPI file that is to be used. for now, keep the java fallback but lets revisit that once the jenkins workflow is functioning properly

File: apps/bfd-data-npi/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+BFD NPI to Organization Name Lookup
+=======================================
+
+This maven project produces a jar file that provides functionality to retrieve organization names given an NPI as input. This lookup is backed by a data file stored in the jar files as a resource.
+
+The data file is constructed from
```
Comment (christrotteradhoc) [inline apps/bfd-data-npi/README.md:6]: suggest: `The data file is constructed from a pipe delimited file an example of which is given below`

Also, lets provide a link to that NPI landing page and the most recent CSV file for reference

File: apps/bfd-data-npi/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+BFD NPI to Organization Name Lookup
+=======================================
+
+This maven project produces a jar file that provides functionality to retrieve organization names given an NPI as input. This lookup is backed by a data file stored in the jar files as a resource.
+
+The data file is constructed from 
+NPI|Entity Type Code|Replacement NPI|Employer Identification Number (EIN)|Provider Organization Name (Legal Business Name)Provider Last Name (Legal Name)|Provider First Name|
```
Comment (christrotteradhoc) [inline apps/bfd-data-npi/README.md:7]: I think there is a pipe character missing between the org nam and the provider last name

File: apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/utility/DataUtilityCommons.java [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    } catch (IOException e) {
+      throw new IllegalStateException("Cannot create temporary directory.");
+    }
+
+    // If the output file isn't already there, go build it.
+    Path convertedNpiDataFile = outputPath.resolve(npiFile);
+    try {
+      buildNPIResource(convertedNpiDataFile, downloadUrl, tempDir);
+    } catch (IOException exception) {
+      LOGGER.error("NPI data file could not be read.  Error:", exception);
```
Comment (christrotteradhoc) [inline apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/utility/DataUtilityCommons.java:71]: same question on flow control -- we log an error but don't re-throw. How is the caller supposed to handle this?

File: apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/lookup/NPIOrgLookup.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+   *
+   * @param npiNumber - npiNumber value in claim records
+   * @return the npi org data display string
+   */
+  public Optional<String> retrieveNPIOrgDisplay(Optional<String> npiNumber) {
+    /*
+     * Handle bad data (e.g. our random test data) if npiNumber is empty
+     */
+    if (!npiNumber.isPresent() && npiOrgHashMap.isEmpty()) {
+      return null;
```
[the remark this answers is not in the document]
Comment (cbrunefearless) [reply apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/lookup/NPIOrgLookup.java:82]: CHanged to Optional Empty..

File: apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/utility/DataUtilityCommons.java (hunk 4 above)
Comment (cbrunefearless) [reply apps/bfd-data-npi/src/main/java/gov/cms/bfd/data/npi/utility/DataUtilityCommons.java:71]: rethrowing error

File: apps/bfd-data-npi/README.md (hunk 3 above)
Comment (cbrunefearless) [reply apps/bfd-data-npi/README.md]: done

File: apps/bfd-data-npi/README.md (hunk 2 above)
Comment (cbrunefearless) [reply apps/bfd-data-npi/README.md:12]: done

**Review by christrotteradhoc — APPROVED** [review]

**Review by mackec1 — APPROVED** [review]
I thought I had approved yesterday...so I'll leave same comment. Looks good...probably a couple of places where you didn't need to be so verbose on some things (i.e., checking args count GT...if you don't ref/use, say args[2] then you don' t to even worry about the args count)

**Pull request merged**